src/clj/fluree/db/flake/reasoner.cljc

```diff
@@ -235,21 +258,47 @@
             (recur r db**))
           db*)))))
 
+(defn find-duplicate-ids
+  [raw-rules]
+  (let [rule-ids (map first raw-rules)]
+    (->> rule-ids
+         frequencies
+         (filter #(< 1 (val %))))))
+
+(defn deduplicate-raw-rules
+  "Given a list of reasoning rules, identifies rules with duplicate ids and renames them using
+  indexes."
+  [raw-rules]
+  (let [duplicate-id-frequencies (find-duplicate-ids raw-rules)]
+    (reduce (fn [rules [duplicate-id occurances]]
+              (let [grouped-rules (group-by #(= duplicate-id (first %)) rules)]
+                (loop [suffix occurances
+                       rules-to-update (get grouped-rules true)
+                       updated-rules-list (get grouped-rules false)]
+                  (if (empty? rules-to-update)
+                    updated-rules-list
+                    (let [updated-rule [(str duplicate-id suffix) (last (first rules-to-update))]]
+                      (recur (dec suffix) (rest rules-to-update) (conj updated-rules-list updated-rule)))))))
+            raw-rules duplicate-id-frequencies)))
+
 (defn reason
-  [db methods rules-graph fuel-tracker reasoner-max]
+  [db methods rule-sources fuel-tracker reasoner-max]
   (go-try
-    (let [db*             (update db :reasoner #(into methods %))
-          tx-state        (flake.transact/->tx-state :db db*)
-          inserts         (atom nil)
+    (let [db*                (update db :reasoner #(into methods %))
+          tx-state           (flake.transact/->tx-state :db db*)
+          inserts            (atom nil)
           ;; TODO - rules can be processed in parallel
-          raw-rules       (<? (all-rules methods db* inserts rules-graph))
-          _               (log/debug "Reasoner - extracted rules: " raw-rules)
-          reasoning-rules (-> raw-rules
-                              resolve/rules->graph
-                              add-rule-dependencies)
-          db**            (if-let [inserts* @inserts]
-                            (<? (process-inserts db* fuel-tracker inserts*))
-                            db*)]
+          raw-rules          (<? (all-rules methods db* inserts rule-sources))
+          _                  (log/debug "Reasoner - extracted rules: " raw-rules)
+          duplicate-id-freqs (find-duplicate-ids raw-rules)
+          deduplicated-rules (when (not (empty? duplicate-id-freqs))
+                               (log/error "Duplicate ids detected. Some rules will be overwritten:" (apply str (map first duplicate-id-freqs))))
```

Comment:
This is not an error and should not be logged as such. At most this should be logged at the warn level, but I think it could reasonably be logged at info or debug.